### DISCUSSION TITLE
2.5 Debugging Code

### DISCUSSION THREAD

**Discussion by bikash119** [Lightning-AI/dl-fundamentals#106] [Lecture videos and quizzes] [answered]
@rasbt showed a neat trick to debug code in notebooks. Thank you @rasbt . The issue that I use to face with the approach is if I have to debug a code inside a loop the debugger won't come out until and unless the loop completes.
```
class BankNoteDataSet(Dataset):
  def __init__(self,csv_file):
    super().__init__()
    self.df = pd.read_csv(csv_file)

  def __getitem__(self, idx) :
    if torch.is_tensor(idx):
      idx = idx.tolist()
    input_features = self.df.iloc[idx,0:4].to_numpy(dtype='float32')
    output_label = self.df.iloc[idx,4]
    # # Debug statement
    # print(f"Entry : {idx} Input Features :{input_features}, output_labels=f{output_label}")
    try:
      input_features = torch.tensor(input_features)
      output_label = torch.tensor(float(output_label))
    except:
      import pdb; pdb.post_mortem()
    # # Debug Statement
    # print(f"##### Input Features :{input_features}, output_labels=f{output_label}")
    return input_features,output_label

  def __len__(self):
    return len(self.df)
```
Since `__getitem__` gets invoked during training loop for every row of training dataset, the debugger would stop for each row which results in exeception. In most of the situations, the failure reason of the rows would be same. In those scenarios, the debugger won't let us move out or stop debugging. If we try to stop the execution of cell by clicking on the stop button on the left of cell, a new pdb textbox would appear. The only way to come out of the debugger is to do restart session of the notebook.

To get over this, we can use `%debug` magic command just after the error throwing cell. It performs the same way as pdb but gives more control on when you want to get out of the debugger. Additionally, we don't have to include any debugging code into our code. I learnt this from [Using a debugger in notebook](https://youtu.be/rSPyvPw0p9k?si=rt4YKPoflN28rl_K) by [Sylvain](https://twitter.com/GuggerSylvain)

**Comment by rasbt** [comment, marked as answer]
This is a nice one, thanks for suggesting @bikash119 !!
